core/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -642,6 +635,11 @@ public void process(MetaData metaData, @Nullable MappingMetaData mappingMd, bool
         }
     }
 
+    /* resolve the routing if needed */
+    public void resolveRouting(MetaData metaData) {
```

Comment:
The corresponding method resolveAndValidateRouting for deletes is a static method on TransportDeleteAction. For consistency, we should either make this one a static method  on TransportIndexAction or move the one in TransportDeleteAction as instance method on DeleteRequest (my preference). Same for resolveAndValidateRouting in TransportUpdateAction.